processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -1039,98 +1038,55 @@ public void set(Offset currOffset)
 
-      final DimensionSelector selector = cursor.makeDimensionSelector(
-          new DefaultDimensionSpec(dimension, dimension)
+      final DimensionQueryHelper queryHelper = DimensionHandlerUtils.makeBaseQueryHelper(
+          dimension,
+          cursor.getColumnCapabilities(dimension),
+          Lists.<String>newArrayList(storageAdapter.getAvailableDimensions())
       );
-
-      // if matching against null, rows with size 0 should also match
-      final boolean matchNull = Strings.isNullOrEmpty(value);
-
-      final int id = selector.lookupId(value);
-      if (id < 0) {
-        return new BooleanValueMatcher(false);
-      } else {
-        return new ValueMatcher()
-        {
-          @Override
-          public boolean matches()
-          {
-            IndexedInts row = selector.getRow();
-            if (row.size() == 0) {
-              return matchNull;
-            }
-            for (int i = 0; i < row.size(); i++) {
-              if (row.get(i) == id) {
-                return true;
-              }
-            }
-            return false;
-          }
-        };
-      }
+      return queryHelper.getValueMatcher(cursor, value);
     }
 
     @Override
     public ValueMatcher makeValueMatcher(String dimension, final DruidPredicateFactory predicateFactory)
     {
-      ValueType type = getTypeForDimension(dimension);
-      switch (type) {
-        case LONG:
+      if (dimension.equals(Column.TIME_COLUMN_NAME) || availableMetrics.contains(dimension)) {
+        if (getTypeForDimension(dimension) == ValueType.LONG) {
           return makeLongValueMatcher(dimension, predicateFactory.makeLongPredicate());
-        case STRING:
-          return makeStringValueMatcher(dimension, predicateFactory.makeStringPredicate());
-        default:
-          return new BooleanValueMatcher(predicateFactory.makeStringPredicate().apply(null));
+        }
       }
-    }
 
-    private ValueMatcher makeStringValueMatcher(String dimension, final Predicate<String> predicate)
-    {
-      final DimensionSelector selector = cursor.makeDimensionSelector(
-          new DefaultDimensionSpec(dimension, dimension)
+      final DimensionQueryHelper queryHelper = DimensionHandlerUtils.makeBaseQueryHelper(
+          dimension,
+          cursor.getColumnCapabilities(dimension),
+          Lists.<String>newArrayList(storageAdapter.getAvailableDimensions())
       );
-
-      return new ValueMatcher()
-      {
-        final boolean matchNull = predicate.apply(null);
-
-        @Override
-        public boolean matches()
-        {
-          IndexedInts row = selector.getRow();
-          if (row.size() == 0) {
-            return matchNull;
-          }
-          for (int i = 0; i < row.size(); i++) {
-            if (predicate.apply(selector.lookupName(row.get(i)))) {
-              return true;
-            }
-          }
-          return false;
-        }
-      };
+      return queryHelper.getValueMatcher(cursor, predicateFactory);
```

Comment:
This is kind of sketchy, since `makeBaseQueryHelper` is presumably intended to return null in some situations. It looks like in any current situation, it _won't_ return null (due to the capabilities substitution) but it does seem intended to possibly return null.

Please either specify that `makeBaseQueryHelper` will never return null, or else put in a null check here.